### DISCUSSION TITLE
Sequencing and scheduling tools in cl-collider

### DISCUSSION THREAD

**Discussion by kflak** [byulparan/cl-collider#151] [Q&A] [answered]
Realized that my question over at the quarks discussion page probably should be a thread of its own... I am trying to understand the different ways of scheduling and sequencing in cl-collider. In sclang there are tasks and routines, as well as the *Clock.sched types of functions. It would be great to get an overview (with examples) of the different ways of doing similar things in cl.

**Comment by kflak** [comment]
On the sbcl side of things, I found `schedule-timer`, which should do the things I want to do, but this is not very portable, as it is only working on sbcl:

```common-lisp
(schedule-timer (make-timer (lambda ()
                              (print "hello!")
                              (force-output))
                            :thread t)
                2)
```
Setting the `:thread` paramer to `t` ensures that I can run the timer within emacs/sly without problem.

**Reply by cianoc** [reply]
Have you seen this:
https://github.com/defaultxr/cl-collider-tutorial/blob/master/06-sequencing.org

For your purposes you probably want to look at the functions available in package.lisp (these are the functions that are exported), particularly:
```cl
    #:sched-ahead
    #:latency

    #:now
    #:callback
    #:quant
    #:set-clock
    #:clock-bpm
    #:clock-beats
    #:clock-add
    #:clock-quant
    #:clock-dur
    #:clock-clear
    #:at-beat
    #:at-task
```

Those should give you what you need - and look similar enough to SuperCollider's scheduling that it should be relatively easy to port I would think.

**Reply by kflak** [reply]
Hi @cianoc! 

Thanks :-) I looked into these, and as far as I understand they are related to scheduling on the server. I haven't found anything in the package that emulates the language-side scheduling in sclang, where I can trigger all kinds of things that are not necessarily scsynth-related. To get around this I made this package: https://codeberg.org/kflak/qp, a port of some of the basic functionality of the CuePlayer quark. I also figured out that I can implement `fork` like this: 
```common-lisp
(defmacro fork (&body body)
  `(bt:make-thread
    (lambda ()
      ,@body)))
```
Then I can do stuff like this:
```
(fork 
      (print (+ 1 1))
      (force-output)
      (sleep 2)
      (force-output)
      (print (+ 2 2)))
```
Only works on sbcl, though...

**Comment by cianoc** [comment, marked as answer]
Yes callback can do that:
```cl
(progn
  (print (+ 1 1))
  (force-output)
  (callback (+ (now) 2) (lambda () (print (+ 2 2)) (force-output))))
```

of if you have a function that you want to use, you could also write that a little bit more elegantly:
```cl
(progn
  (+ 1 1)
  (callback (+ (now) 2) #'+ 1 1))
```

Obviously you might want to wrap that in a macro, to remove the need for constantly writing (+ (now) time) - but the basic mechanism is there.

For example you could have something like the following:
```cl
(schedule
  (0 #'+ 1 1)
  (1 (lambda (x y) (+ x y)) 3 4))
```

Things to bear in mind. There will be latency, so if you have 300ms latency, these occur 300ms earlier (this would also be true if you were using sclang). So you might want to adjust for that by adding latency to now. Also this is using time, so you'd also need some way to convert if you're using beats.

**Comment by cianoc** [comment]
You can also use tempo clock to get a scheduler that works in beats.

**Comment by kflak** [comment]
Very nice, thanks a lot! Is there a way to get around the limitation that the server needs to be up and running for the scheduling to work?

**Comment by byulparan** [comment]
There is a method like the following.
A server is required, even if only as a dummy, because some information needs to be retrieved from the server.
Booting the server is not necessary.
Since an unbooted server was not part of the use case I had in mind, these symbols are not exported.
Symbols exported like `clock-add` are APIs for the scheduler included in the server (they are just very simple wrappers), so you will need to create a separate tempo clock or scheduler.

```lisp
(in-package :sc-user)

(setf *s* (make-external-server "localhost" :port 60111))


(defvar *scheduler* (make-instance 'sc::tempo-clock :server *s*))
(sc::tempo-clock-run *scheduler*)

(sc::tempo-clock-add *scheduler* (+ 1 (sc::tempo-clock-beats *scheduler*))
		     (lambda () (uiop:println "hello world")))

```

**Comment by kflak** [comment]
Cool! I don't expect to use the scheduler without the server running, though, but I could imagine a situation where I would want to control graphics or something else without having to fire up the server first, in which case this would come in handy.
